### ISSUE TITLE
Eval in test mode gives constant error value 0.5  

### ISSUE THREAD

**Issue by ncalik** [vlfeat/autonn#29]
Hi again, I'm sorry to keep you busy
Firstly, I have done the dagg2autonn translating work with the information you provided. Thank you for that. Now I'm working with resnet. training is well but in test mode error gives "0.5" constant value for every iteration. When I use forward mode, testing process works... Where is the mistake in test mode?
And also in test mode when I use training data for test, I get same results... 

net.eval({'data', images, 'label', labels}, 'test') ;
net.eval({'data', images, 'label', labels}, 'forward') ;
loss1 : objective
loss2 : error

![adsiz](https://user-images.githubusercontent.com/17113517/36282475-57018d66-12b2-11e8-8222-02a1e2450ab0.jpg)

**Comment by jotaf98** [comment]
I think "dag2autonn" (or similar) was an old name for this same function; it should be Layer.fromDagNN. Are you sure you're using the most recent version?

If that doesn't solve it, you can put a breakpoint during the validation phase, print the "net" object into the console, and click "Display variables". You can see the sizes, min and max of each tensor, and see if any of them looks weird (e.g. something is constant [same min and max value] when it shouldn't be). If that doesn't help I can look into the converter tomorrow to see if something is wrong.

**Comment by ncalik** [comment]
Hi dag2autonn is not function. I gave this name to the converting process :)
I also trained SE-BN-Net obtained from Albanie's site. By using "forward" mode, it works. 
Here is my codes and network architecture.  
%========================================================%
out   = 2;
net   = dagnn.DagNN.loadobj('SE-BN-Inception-mcn.mat');
layer_fc = net.layers(net.getLayerIndex('classifier')) ; 
params      = layer_fc.params  ;
h   = layer_fc.block.size(1) ;
w   = layer_fc.block.size(2) ;
in  = layer_fc.block.size(3) ;
sc           = sqrt(2/(h*w*out)) ;
filters     = randn(h, w, in, out, 'single')*sc ;
biases    = zeros(out, 1, 'single') ;

net.params(net.getParamIndex(params{1})).value = filters ;
net.params(net.getParamIndex(params{2})).value = biases ;
%---------------------------------------------------------%
net.removeLayer(net.layers(end).name) ;
nameX = net.layers(end).name;

net.addLayer('loss1', ...
dagnn.Loss('loss', 'softmaxlog') ,...
{nameX, 'label'}, ...
'obj') ;

net.addLayer('loss2', ...
dagnn.Loss('loss', 'classerror') ,...
{nameX, 'label'}, ...
'err') ;

outputs = Layer.fromDagNN(net);
lyr =  outputs{:} ;
lyr.sequentialNames;
netx = Net(lyr ) ;
%========================================================%
Is it possible that  in test mode, the label block I added as an input is not functioning

![adsiz2](https://user-images.githubusercontent.com/17113517/36294842-7f6ae774-12f4-11e8-85fb-4b80c96c24ed.jpg)
[SE-BN-Net.pdf](https://github.com/vlfeat/autonn/files/1730312/SE-BN-Net.pdf)

**Comment by jotaf98** [comment]
Yes, that's possible. This is how I would do it (untested) :

```
net = dagnn.DagNN.loadobj('SE-BN-Inception-mcn.mat');
outputs = Layer.fromDagNN(net);

pred = outputs{1}.find('classifier');  % easier than working with the DAG
labels = Input('name', 'labels');
objective = vl_nnloss(pred, labels, 'loss', 'softmaxlog') ;
error = vl_nnloss(pred, labels, 'loss', 'classerror') ;

netx = Net(objective, error) ;  % sequentialNames is already done automatically when compiling
```

Also, make sure the layer you're getting is *before* the softmax. Otherwise, you'll be training with 2 composed softmaxes: the first that comes with the network, and the second that is part of the loss (remember that for numerical stability the log loss already includes the softmax).

EDIT: Added losses

**Comment by ncalik** [comment]
Thank you for replying. After applying your code, I've got an error : 

**No var with specified name 'data'.
Error in Net/eval (line 55)
    var = net.getVarIndex(inputs{i}) ;
Error in Untitled8 (line 150)
      net.eval({'data', images, 'labels', labels}) ;**

When I looked vars in netx there is only labels   as "input" : 
![image](https://user-images.githubusercontent.com/17113517/36330694-c99ceca2-137b-11e8-88c2-fe6ecbc670d6.png)

Is it possible that "data" input could not move from dagg to autonn properly?
but in other case, by using outputs{1}.find('conv1_7x7_s2'); I can see "data" is input of this conv layer..
 
thanks for this important warning about softmax, a point I should not miss...

Also, while I was dealing with "forward" mode , I noticed that "classifier" layer results are changing acording to the number of images I gave for testing. What cause this?

**Comment by ncalik** [comment]
I solved the problem... It tooks me 2 hours :) 
After these code : 

**labels = Input('name', 'labels');
objective = vl_nnloss(pred, labels, 'loss', 'softmaxlog') ;
error = vl_nnloss(pred, labels, 'loss', 'classerror') ;**

objective.inputs{1} is "cell" type. I think that Net() compiler search "Layer". It's true or not I dont know!!
in objective, subLayer is in objective.inputs{1} cell. When I changed code like this : 

`objective.inputs{1} = objective.inputs{1}{1};
error.inputs{1}     = error.inputs{1}{1};` 
Net() compiled successfully... now it's working. Your code is very helpfull for fine-tuning 

But as result noting change in validation process in "test" mode :)

![image](https://user-images.githubusercontent.com/17113517/36333047-d4c026a8-1385-11e8-8981-489cfab92c6e.png)


When I change "test" to "forward" : 
![image](https://user-images.githubusercontent.com/17113517/36333193-80f57a22-1386-11e8-8f11-6a17247d7d3e.png)

But at this time I cant get robust value from "classifier" layer.

**Comment by jotaf98** [comment]
It's great that you managed to solve it! :)

I wish I could dig deeper in this but unfortunately I can't for the next couple of days. I've seen this before though, it is usually due to batch norm. Batch norm uses different moments at training and test time. The training moments are obtained from the batch itself. The test moments, though, are fixed; they are a running average of the training moments. (This is pretty standard.) The problem you're seeing must be because the moments used at test time are somehow incorrect. So if you look into that you'll find the problem. The training is still ok. I'll try to look into it when I get the chance but as I said it will take some time.

**Comment by ncalik** [comment]
Hi,
now I used cnn_train_autonn instead of hand-made updating. I compared results in autonn-dagnn results using resnet-50 
![autonn-dagnn pg](https://user-images.githubusercontent.com/17113517/36417429-ff1dcf18-163c-11e8-96ea-639a06b06024.jpg)
rigth is autonn left is dagnn

In dagnn, val err can change but in auto, its constant again...

**Comment by jotaf98** [comment]
Hi, can you try the following?

There's a new easier interface to training that replaces cnn_train_autonn, I've been working with it on another branch for several months, it's pretty stable and now I've merged it. (The old training function is still available.) So can you update to the latest version, then try:

```
cifar_example('model',models.ResNet('numClasses',10))
```

This should start training a ResNet-50 from scratch. It's just to test things. Then to move on to a pre-trained network:

```
prediction = models.ResNet('pretrained', true);
assert(isequal(prediction.func, @vl_nnconv));  % ensure last layer is conv
prediction.inputs{2}.value(:,:,:,11:end) = [];  % leave only 10 output classes in weights
prediction.inputs{3}.value(11:end) = [];  % leave only 10 output classes in biases
cifar_example('model',prediction)
```

I'm not sure that will attain a low training error without tuning the learning rate, but it should run OK. Let me know if you observe the same behaviour in this new setting. It's going to be easier to debug than the old function. If you want to try it on a custom dataset have a look at examples/cnn, there are 2 examples for that (one with the ImageFolder class, the other with an IMDB-type structure).

**Comment by ncalik** [comment]
The work you're doing is really great. You effectively blend Keras-like code style with Matlab environments. I hope that more people will notice your work and use autonn.
I'll try new interface as soon as possible, then inform you..

**Comment by jotaf98** [comment]
Thanks, I really appreciate that! :)

**Comment by ncalik** [comment]
Hi, after updating autonn files,it works really cool. Both the new codes you write and my old codes do good work. Also there is no problem using cnn_train_autonn !! I tried models.Resnet and custom dataset and they gave substantial results. Many thanks !!... 👍 👍 

There is a quest. about dataset preparing. 
Should we always prepare folders in this style : 
<main_set> / train /<1><2> ... <Ncls> 
<main_set> / val / <1><2> ... <Ncls> 
or can we use only  
<main_set>/<1><2>...<Ncls>  where <.> is class folder of jpg images? 

![new](https://user-images.githubusercontent.com/17113517/36529764-59e8f4a2-17ca-11e8-9563-aa2fe9316bd6.png)
here is pretrained resnet (obtained from old codes mentioned above) 
![new2](https://user-images.githubusercontent.com/17113517/36529830-8ccedac6-17ca-11e8-8544-304054da6359.png)
this is the SE-BN-Net

**Comment by jotaf98** [comment]
Awesome!!

Well there's no restriction since ImageFolder has no concept of labels. It will list all the filenames though, so you can then loop over the names and decide for yourself which label to assign to which. I struggled a bit with this design decision but in the end I think it's more flexible. You can see how it's used in the "image_folder_example.m". Happy coding!

EDIT: If you're using cnn_train_autonn (which means your dataset is a IMDB) then also it's completely up to you. There's no hard rule, it's however you prefer to keep things organized ;)
